nssp/delphi_nssp/patch.py

```diff
@@ -0,0 +1,155 @@
+    if valid_config:
+        logger.info("Good patch configuration.")
+        return True
+    logger.info("Bad patch configuration.")
+    return False
+
+
+def patch():
+    """
+    Run the nssp indicator for a range of issue dates.
+
+    The range of issue dates is specified in params.json using the following keys:
+    - "patch": Only used for patching data
+        - "source_backup_credentials": (optional) dict, credentials to log in to
+        server where historical source data is backed up.
+        if "source_dir" doesn't exist or has no files in it, we download source data to source_dir before running patch.
+        else, we assume all needed source files are already in source_dir.
+            - "host": str, hostname of the server where source data is backed up
+            - "user": str, username to log in to the server
+            - "path": str, path to the directory containing backup csv files
+        - "start_date": str, YYYY-MM-DD format, first issue date
+        - "end_date": str, YYYY-MM-DD format, last issue date
+        - "patch_dir": str, directory to write all issues output
+        - "source_dir": str, directory to read source data from.
+    """
+    params = read_params()
+    logger = get_structured_logger("delphi_nssp.patch", filename=params["common"]["log_filename"])
+    if not good_patch_config(params, logger):
+        sys.exit(1)
+
+    source_dir = params["patch"]["source_dir"]
+    downloaded_source = False
+    if not path.isdir(source_dir) or not listdir(source_dir):
+        get_source_data(params, logger)
+        downloaded_source = True
+
+    start_issue = datetime.strptime(params["patch"]["start_issue"], "%Y-%m-%d")
+    end_issue = datetime.strptime(params["patch"]["end_issue"], "%Y-%m-%d")
+
+    logger.info(start_issue=start_issue.strftime("%Y-%m-%d"))
+    logger.info(end_issue=end_issue.strftime("%Y-%m-%d"))
+    logger.info(source_dir=source_dir)
+    logger.info(patch_dir=params["patch"]["patch_dir"])
+
+    makedirs(params["patch"]["patch_dir"], exist_ok=True)
+
+    current_issue = start_issue
+    while current_issue <= end_issue:
+        logger.info("patching issue", issue_date=current_issue.strftime("%Y-%m-%d"))
+
+        current_issue_source_csv = f"""{source_dir}/{current_issue.strftime("%Y-%m-%d")}.csv"""
+        if not path.isfile(current_issue_source_csv):
+            logger.info("No source data at this path", current_issue_source_csv=current_issue_source_csv)
+            current_issue += timedelta(days=1)
+            continue
+
+        params["patch"]["current_issue"] = current_issue.strftime("%Y-%m-%d")
+
+        # current_issue_date can be different from params["patch"]["current_issue"]
+        # due to weekly cadence of nssp data. For weekly sources, issue dates in our
+        # db matches with first date of epiweek that the reporting date falls in,
+        # rather than reporting date itself.
+        current_issue_date = Week.fromdate(current_issue).startdate()
```

Comment:
issue: Since our `current_issue` `while` loop goes day-by-day, we will get the same `current_issue_date` for 7 `current_issue`s in a row (Sunday-Saturday of the week starting with a given `current_issue_date`). That means that the `run_module` call below will generate the same data.

This can fixed by keeping track of which `current_issue_date`s we've seen before (e.g. in a set) and skipping loop iterations/`current_issue`s where the corresponding `current_issue_date` is in the set.